### ISSUE TITLE
[Bug]: Load collection failed when running `verify_all_collections.py` with error message `failed to get segment 433948783414607873` after standalone pod kill chaos

### ISSUE THREAD

**Issue by zhuwenxing** [milvus-io/milvus#17607]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:master-20220616-53a49a92
- Deployment mode(standalone or cluster): standalone
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus==2.1.0.dev69
- OS(Ubuntu or CentOS):  
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

Load failed when running `verify_all_collections.py`

```
20:47:32  check collection Checker__wAS9PJDA
20:47:32  collection is exist
20:47:32  
20:47:32  Create collection...
20:47:32  
20:47:32  Insert 3000 vectors cost 0.0707 seconds
20:47:32  
20:47:32  Get collection entities...
20:47:32  6130
20:47:32  
20:47:32  Get collection entities cost 3.0165 seconds
20:47:32  
20:47:32  Create index...
20:47:32  
20:47:32  Create index cost 1.0320 seconds
20:47:32  
20:47:32  load collection...
20:47:32  Traceback (most recent call last):
20:47:32    File "scripts/verify_all_collections.py", line 124, in <module>
20:47:32      hello_milvus(collection_name)
20:47:32    File "scripts/verify_all_collections.py", line 74, in hello_milvus
20:47:32      collection.load()
20:47:32    File "/usr/local/lib/python3.7/dist-packages/pymilvus/orm/collection.py", line 450, in load
20:47:32      conn.load_collection(self._name, replica_number=replica_number, timeout=timeout, **kwargs)
20:47:32    File "/usr/local/lib/python3.7/dist-packages/pymilvus/decorators.py", line 61, in handler
20:47:32      raise e
20:47:32    File "/usr/local/lib/python3.7/dist-packages/pymilvus/decorators.py", line 44, in handler
20:47:32      return func(self, *args, **kwargs)
20:47:32    File "/usr/local/lib/python3.7/dist-packages/pymilvus/decorators.py", line 79, in handler
20:47:32      raise e
20:47:32    File "/usr/local/lib/python3.7/dist-packages/pymilvus/decorators.py", line 75, in handler
20:47:32      return func(*args, **kwargs)
20:47:32    File "/usr/local/lib/python3.7/dist-packages/pymilvus/client/grpc_handler.py", line 683, in load_collection
20:47:32      raise MilvusException(response.error_code, response.reason)
20:47:32  pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=, failed to get segment 433948783414607873)>
```

### Expected Behavior

all test cases passed

### Steps To Reproduce

```markdown
see http://10.100.32.144:8080/blue/organizations/jenkins/chaos-test/detail/chaos-test/6555/pipeline
```

### Milvus Log

failed job: http://10.100.32.144:8080/blue/organizations/jenkins/chaos-test/detail/chaos-test/6555/pipeline
log:
[artifacts-standalone-pod-kill-6555-server-logs.tar.gz](https://github.com/milvus-io/milvus/files/8919257/artifacts-standalone-pod-kill-6555-server-logs.1.tar.gz)
 

### Anything else?

many other pod kill chaos test also have this issue,
see 
https://github.com/zhuwenxing/milvus/runs/6918318821?check_suite_focus=true
https://github.com/zhuwenxing/milvus/runs/6918319261?check_suite_focus=true
https://github.com/zhuwenxing/milvus/runs/6918319606?check_suite_focus=true

**Comment by congqixia** [comment]
From the log, the segment `433948783414607873` is marked as dropped. Maybe due to previously compaction before chaos. Might be related to #17567.
Still looking into root cause.

**Comment by congqixia** [comment]
https://github.com/milvus-io/milvus/blob/0f877636821d40f79b17097bbf4824ab9a49924c/internal/querycoord/task.go#L2594-L2609
This function tries to fetch segmentInfo with dropped segment id while datacoord will report error when a segment is marked dropped.
@wayblink @sunby any workaround to fix this problem?

**Comment by wayblink** [comment]
https://github.com/milvus-io/milvus/blob/0f877636821d40f79b17097bbf4824ab9a49924c/internal/datacoord/meta.go#L198-L208
https://github.com/milvus-io/milvus/blob/0f877636821d40f79b17097bbf4824ab9a49924c/internal/datacoord/meta.go#L1056-L1060

It is because meta.GetSegment will call isSegmentHealthy, will return nil for a dropped segment.
I don't know the background of this restriction: `segment.GetState() != commonpb.SegmentState_Dropped`. 
Could you explain it if you remember？@sunby

Considering `isSegmentHealthy` is used in many places. I 'd like to create a new GetAllSegmentInfo RPC which can also return dropped segments. What do you think about it? @congqixia

**Comment by wayblink** [comment]
> https://github.com/milvus-io/milvus/blob/0f877636821d40f79b17097bbf4824ab9a49924c/internal/datacoord/meta.go#L198-L208
> 
> 
> https://github.com/milvus-io/milvus/blob/0f877636821d40f79b17097bbf4824ab9a49924c/internal/datacoord/meta.go#L1056-L1060
> 
> It is because meta.GetSegment will call isSegmentHealthy, will return nil for a dropped segment. I don't know the background of this restriction: `segment.GetState() != commonpb.SegmentState_Dropped`. Could you explain it if you remember？@sunby
> 
> Considering `isSegmentHealthy` is used in many places. I 'd like to create a new GetAllSegmentInfo RPC which can also return dropped segments. What do you think about it? @congqixia

Or add an option field in GetSegmentInfoRequest

**Comment by congqixia** [comment]
@wayblink `isSegmentHealthy` check is required since `GetSegmentInfo` provides service for  `GetPersistentSegmentInfo` which shall return only `healthy` segments.
Add option or new API seem doable to me. However, seperate `GetRecoveryInfo` and `GetSegmentInfo` has some risk that one segment is cleaned up by GC. So do we need add reference lock here? 
@sunby WDYT

**Comment by wayblink** [comment]
@zhuwenxing A pr is merged to fix this issue. Could you help to verify it, please?

**Comment by zhuwenxing** [comment]
For some other pod kill chaos tests, the error message of load failure changed to 
```
pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=collection hello_milvus has not been loaded to memory or load failed)>
```

see: 
etcd:https://github.com/zhuwenxing/milvus/runs/6936879067?check_suite_focus=true
rootcoord: https://github.com/zhuwenxing/milvus/runs/6936878971?check_suite_focus=true
proxy: https://github.com/zhuwenxing/milvus/runs/6936878474?check_suite_focus=true
datacoord: https://github.com/zhuwenxing/milvus/runs/6936877934?check_suite_focus=true

**Comment by zhuwenxing** [comment]
The issue https://github.com/milvus-io/milvus/issues/17607#issuecomment-1158959357 is not reproduced today but it seems there is no fix for it.

**Comment by wayblink** [comment]
> The issue [#17607 (comment)](https://github.com/milvus-io/milvus/issues/17607#issuecomment-1158959357) is not reproduced today but it seems there is no fix for it.

You mean the load_collection error above still exists? 

> For some other pod kill chaos tests, the error message of load failure changed to
> 
> ```
> pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=collection hello_milvus has not been loaded to memory or load failed)>
> ```
> 
> see: etcd:https://github.com/zhuwenxing/milvus/runs/6936879067?check_suite_focus=true rootcoord: https://github.com/zhuwenxing/milvus/runs/6936878971?check_suite_focus=true proxy: https://github.com/zhuwenxing/milvus/runs/6936878474?check_suite_focus=true datacoord: https://github.com/zhuwenxing/milvus/runs/6936877934?check_suite_focus=true

It seems like another issue? error is about load_collection instead of get_segment_info.

**Comment by wayblink** [comment]
> The issue [#17607 (comment)](https://github.com/milvus-io/milvus/issues/17607#issuecomment-1158959357) is not reproduced today but it seems there is no fix for it.

You mean the load_collection error still exist?

**Comment by wayblink** [comment]
> The issue [#17607 (comment)](https://github.com/milvus-io/milvus/issues/17607#issuecomment-1158959357) is not reproduced today but it seems there is no fix for it.

This one: https://github.com/milvus-io/milvus/pull/17617
So the load_collection error didn't occur today?

**Comment by congqixia** [comment]
@zhuwenxing I think @wayblink's #17617 and letian's #17619 shall fix this problem.

**Comment by zhuwenxing** [comment]
> > The issue [#17607 (comment)](https://github.com/milvus-io/milvus/issues/17607#issuecomment-1158959357) is not reproduced today but it seems there is no fix for it.
> 
> This one: #17617 So the load_collection error didn't occur today?

Yes, now the error message of load is not `failed to get segment xxx`, but changed to `has not been loaded to memory or load failed`

a new failed job
https://github.com/milvus-io/milvus/runs/6956973459?check_suite_focus=true

**Comment by wayblink** [comment]
https://github.com/milvus-io/milvus/suites/6997534572/artifacts/274320003
datanode：
[2022/06/19 18:55:01.014 +00:00] [DEBUG] [segment_loader.go:171] ["start to load segments in parallel"] [segmentNum=8] [concurrencyLevel=2]
[2022/06/19 18:55:01.014 +00:00] [DEBUG] [parallel.go:55] [loadSegmentFunc] [total=8]
[2022/06/19 18:55:01.014 +00:00] [DEBUG] [parallel.go:56] [loadSegmentFunc] [nPerBatch=4]
[2022/06/19 18:55:01.014 +00:00] [DEBUG] [parallel.go:108] [loadSegmentFunc] [NumOfGoRoutines=2]
[2022/06/19 18:55:01.014 +00:00] [INFO] [segment_loader.go:205] ["start loading segment data into memory"] [collectionID=434022443741023553] [partitionID=434022443741023554] [segmentID=434022617569296385]
[2022/06/19 18:55:01.014 +00:00] [INFO] [segment_loader.go:205] ["start loading segment data into memory"] [collectionID=434022443741023553] [partitionID=434022443741023554] [segmentID=434022443886575617]
[2022/06/19 18:55:01.022 +00:00] [INFO] [segment.go:695] ["load field done"] [fieldID=1] ["row count"=1497] [segmentID=434022617569296385]
[2022/06/19 18:55:01.036 +00:00] [INFO] [segment.go:695] ["load field done"] [fieldID=101] ["row count"=1497] [segmentID=434022617569296385]
[2022/06/19 18:55:01.036 +00:00] [DEBUG] [segment_loader.go:447] ["load index file"] [path=file/index_files/434022444804603906/1/434022443741023554/434022443886575617/IVF]
[2022/06/19 18:55:01.041 +00:00] [WARN] [minio_chunk_manager.go:194] ["failed to read object"] [path=file/index_files/434022444804603906/1/434022443741023554/434022443886575617/IVF] [error="The specified key does not exist."]
[2022/06/19 18:55:01.042 +00:00] [INFO] [segment.go:695] ["load field done"] [fieldID=100] ["row count"=1497] [segmentID=434022617569296385]
[2022/06/19 18:55:01.053 +00:00] [INFO] [segment.go:695] ["load field done"] [fieldID=102] ["row count"=1497] [segmentID=434022617569296385]
[2022/06/19 18:55:01.128 +00:00] [INFO] [segment.go:695] ["load field done"] [fieldID=0] ["row count"=1497] [segmentID=434022617569296385]
[2022/06/19 18:55:01.132 +00:00] [ERROR] [segment_loader.go:155] ["load segment failed when load data into memory"] [collectionID=434022443741023553] [partitionID=434022443741023554] [segmentID=434022443886575617] ["segment type"=3] [error="The specified key does not exist."] [stack="github.com/milvus-io/milvus/internal/querynode.(*segmentLoader).LoadSegment.func2\n\t/go/src/github.com/milvus-io/milvus/internal/querynode/segment_loader.go:155\ngithub.com/milvus-io/milvus/internal/util/funcutil.ProcessFuncParallel.func3\n\t/go/src/github.com/milvus-io/milvus/internal/util/funcutil/parallel.go:85"]
[2022/06/19 18:55:01.134 +00:00] [ERROR] [parallel.go:87] [loadSegmentFunc] [error="The specified key does not exist."] [idx=4] [stack="github.com/milvus-io/milvus/internal/util/funcutil.ProcessFuncParallel.func3\n\t/go/src/github.com/milvus-io/milvus/internal/util/funcutil/parallel.go:87"]
[2022/06/19 18:55:01.136 +00:00] [INFO] [segment.go:695] ["load field done"] [fieldID=103] ["row count"=1497] [segmentID=434022617569296385]
[2022/06/19 18:55:01.137 +00:00] [INFO] [segment_loader.go:358] ["log field binlogs done"] [collection=434022443741023553] [segment=434022617569296385] [fields="[{\"fieldID\":1,\"binlogs\":[{\"entries_num\":1497,\"log_path\":\"file/insert_log/434022443741023553/434022443741023554/434022617569296385/1/434022618196344838\",\"log_size\":543}]},{\"fieldID\":100,\"binlogs\":[{\"entries_num\":1497,\"log_path\":\"file/insert_log/434022443741023553/434022443741023554/434022617569296385/100/434022618196344833\",\"log_size\":4864}]},{\"fieldID\":101,\"binlogs\":[{\"entries_num\":1497,\"log_path\":\"file/insert_log/434022443741023553/434022443741023554/434022617569296385/101/434022618196344834\",\"log_size\":6298}]},{\"fieldID\":102,\"binlogs\":[{\"entries_num\":1497,\"log_path\":\"file/insert_log/434022443741023553/434022443741023554/434022617569296385/102/434022618196344835\",\"log_size\":5210}]},{\"fieldID\":103,\"binlogs\":[{\"entries_num\":1497,\"log_path\":\"file/insert_log/434022443741023553/434022443741023554/434022617569296385/103/434022618196344836\",\"log_size\":691282}]},{\"binlogs\":[{\"entries_num\":1497,\"log_path\":\"file/insert_log/434022443741023553/434022443741023554/434022617569296385/0/434022618196344837\",\"log_size\":4188}]}]"]
[2022/06/19 18:55:01.138 +00:00] [DEBUG] [segment_loader.go:259] ["loading bloom filter..."] [segmentID=434022617569296385]
[2022/06/19 18:55:01.163 +00:00] [DEBUG] [segment_loader.go:267] ["loading delta..."] [segmentID=434022617569296385]
[2022/06/19 18:55:01.163 +00:00] [INFO] [segment_loader.go:609] ["there are no delta logs saved with segment, skip loading delete record"] [segmentID=434022617569296385]
[2022/06/19 18:55:01.190 +00:00] [INFO] [segment_loader.go:205] ["start loading segment data into memory"] [collectionID=434022443741023553] [partitionID=434022443741023554] [segmentID=434022443902566401]
[2022/06/19 18:55:01.194 +00:00] [DEBUG] [segment_loader.go:447] ["load index file"] [path=file/index_files/434022444804603908/1/434022443741023554/434022443902566401/IVF]
[2022/06/19 18:55:01.200 +00:00] [WARN] [minio_chunk_manager.go:194] ["failed to read object"] [path=file/index_files/434022444804603908/1/434022443741023554/434022443902566401/IVF] [error="The specified key does not exist."]
[2022/06/19 18:55:01.211 +00:00] [ERROR] [segment_loader.go:155] ["load segment failed when load data into memory"] [collectionID=434022443741023553] [partitionID=434022443741023554] [segmentID=434022443902566401] ["segment type"=3] [error="The specified key does not exist."] [stack="github.com/milvus-io/milvus/internal/querynode.(*segmentLoader).LoadSegment.func2\n\t/go/src/github.com/milvus-io/milvus/internal/querynode/segment_loader.go:155\ngithub.com/milvus-io/milvus/internal/util/funcutil.ProcessFuncParallel.func3\n\t/go/src/github.com/milvus-io/milvus/internal/util/funcutil/parallel.go:85"]
[2022/06/19 18:55:01.211 +00:00] [ERROR] [parallel.go:87] [loadSegmentFunc] [error="The specified key does not exist."] [idx=1] [stack="github.com/milvus-io/milvus/internal/util/funcutil.ProcessFuncParallel.func3\n\t/go/src/github.com/milvus-io/milvus/internal/util/funcutil/parallel.go:87"]
[2022/06/19 18:55:01.211 +00:00] [DEBUG] [parallel.go:51] [loadSegmentFunc] ["time cost"=197.596527ms]

**Comment by wayblink** [comment]
@zhuwenxing @congqixia seems like a new issue, do we need to close this one and open a new one?It seems the key problem is failed to read segment from minio. 
```
[minio_chunk_manager.go:194] ["failed to read object"] [path=file/index_files/434022444804603906/1/434022443741023554/434022443886575617/IVF] [error="The specified key does not exist."]
```

**Comment by zhuwenxing** [comment]
> @zhuwenxing @congqixia seems like a new issue, do we need to close this one and open a new one?
> 
> It seems the key problem is failed to read segment from minio.
> 
> ```
> [minio_chunk_manager.go:194] ["failed to read object"] [path=file/index_files/434022444804603906/1/434022443741023554/434022443886575617/IVF] [error="The specified key does not exist."]
> ```
> 
> Do you know who is familiar with this part?

I prefer to open a new issue

**Comment by zhuwenxing** [comment]
I have opened a new issue to trace the latest load error
see https://github.com/milvus-io/milvus/issues/17645

**Comment by congqixia** [comment]
@wayblink from the log you provide, looks like the the missing file was the index log file.  
@xiaocai2333  FYI

**Comment by xiaocai2333** [comment]
> @wayblink from the log you provide, looks like the the missing file was the index log file. @xiaocai2333 FYI

Yes, this is the index file. When loading Segment, if the segment is compacted, the binlog file will not be deleted due to the reference lock, but the index files will be deleted.

**Comment by xiaocai2333** [comment]
I have a suggestion, when there is reference lock on the segment, the segment can not be compacted. @xiaofan-luan @congqixia WDYT
